src/main/java/spiffe/provider/CertificateUtils.java

```diff
@@ -0,0 +1,184 @@
+        Collection<? extends Certificate> certificates =  getCertificateFactory().generateCertificates(new ByteArrayInputStream(input));
+        return certificates.stream()
+                .map(X509Certificate.class::cast)
+                .collect(Collectors.toSet());
+    }
+
+    /**
+     * Generate a single X509Certificate
+     *
+     * @param input as byte array
+     * @return an instance of X509Certificate
+     * @throws CertificateException
+     */
+    static X509Certificate generateCertificate(byte[] input) throws CertificateException {
+        return (X509Certificate) getCertificateFactory().generateCertificate(new ByteArrayInputStream(input));
+    }
+
+    /**
+     * Generates a PrivateKey from the X509SVIDKey ByteArray
+     *
+     * It uses PKCS8EncodedKeySpec that represents the ASN.1 encoding of a private key
+     *
+     * @param input as byte array
+     * @return
+     * @throws NoSuchAlgorithmException
+     * @throws InvalidKeySpecException
+     */
+    static PrivateKey generatePrivateKey(byte[] input) throws NoSuchAlgorithmException, InvalidKeySpecException {
+        KeyFactory keyFactory = KeyFactory.getInstance(PRIVATE_KEY_ALGORITHM, new BouncyCastleProvider());
+        return keyFactory.generatePrivate(new PKCS8EncodedKeySpec(input));
+    }
+
+    /**
+     * Validate a certificate chain against a set of trusted certificates.
+     *
+     * @param chain    certificate chain
+     * @param trustedCerts
+     * @throws CertificateException
+     */
+    static void validate(X509Certificate[] chain, Set<X509Certificate> trustedCerts) throws CertificateException {
+        PKIXParameters pkixParameters = toPkixParameters(trustedCerts);
+        CertPath certPath = CERTIFICATE_FACTORY.generateCertPath(Arrays.asList(chain));
+        try {
+            CERT_PATH_VALIDATOR.validate(certPath, pkixParameters);
+        } catch (CertPathValidatorException | InvalidAlgorithmParameterException e) {
+            throw new CertificateException(e);
+        }
+    }
+
+    /**
+     * Validates that the SPIFFE ID is present and matches the SPIFFE ID configured in
+     * the java.security property ssl.spiffe.accept
+     *
+     * @param chain an array of X509Certificate that contains the Peer's SVID to be validated
+     * @throws CertificateException when either the certificates doesn't have a SPIFFE ID or the SPIFFE ID is not trusted
+     */
+    static void checkSpiffeId(X509Certificate[] chain) throws CertificateException {
+        Optional<String> spiffeId = getSpiffeId(chain[0]);
+        if (!spiffeId.isPresent()) {
+            throw new CertificateException("SPIFFE ID not found in the certificate");
+        }
+
+        String acceptedSpiffeId = Security.getProperty(SSL_SPIFFE_ACCEPT_PROPERTY);
+        if (!StringUtils.equals(spiffeId.get(), acceptedSpiffeId)) {
+            String errorMessage = String.format("SPIFFE ID %s is not trusted", spiffeId.get());
```

Comment:
nit: message should probably read `is not authorized` instead of `is not trusted`, since the latter implies that the certificate was signed by an unknown authority

Reply:
done